AmplifyPlugins/API/AWSAPICategoryPlugin/Configuration/AWSAPICategoryPluginConfiguration.swift

```diff
@@ -34,15 +60,82 @@ public struct AWSAPICategoryPluginConfiguration {
+                                                     authService: authService)
+
+        self.init(endpoints: endpoints,
+                  interceptors: interceptors,
+                  apiAuthProviderFactory: apiAuthProviderFactory,
+                  authService: authService)
+
+    }
+
+    /// Registers an interceptor for the provided API endpoint
+    /// - Parameter interceptor: operation interceptor used to decorate API requests
+    /// - Parameter toEndpoint: API endpoint name
+    mutating func addInterceptor(_ interceptor: URLRequestInterceptor,
+                                 toEndpoint apiName: APIEndpointName) {
+        interceptors[apiName]?.addInterceptor(interceptor)
+    }
+
+    /// Returns all the interceptors registered for `apiName` API endpoint
+    /// - Parameter apiName: API endpoint name
+    /// - Returns: request interceptors
+    internal func interceptorsForEndpoint(named apiName: APIEndpointName?) -> [URLRequestInterceptor] {
+        guard let apiName = apiName, let interceptorsConfig = interceptors[apiName] else {
+            return []
+        }
+        return interceptorsConfig.interceptors
+    }
+
+    /// Returns or create interceptors for the provided endpointConfig
+    /// - Parameters:
+    ///   - endpointConfig: endpoint configuration
+    ///   - authType: if provided override the registered auth interceptor
+    /// - Throws: PluginConfigurationError in case of failure building an instance of AWSAuthorizationConfiguration
+    /// - Returns: An array of URLRequestInterceptor
+    internal func interceptorsForEndpoint(withConfig endpointConfig: EndpointConfig,
+                                          authType: AWSAuthorizationType? = nil) throws -> [URLRequestInterceptor] {
+
+        guard let authType = authType, let apiAuthProviderFactory = self.apiAuthProviderFactory else {
+            return interceptorsForEndpoint(named: endpointConfig.name)
+        }
+
+        var config = EndpointInterceptorsConfig(endpointName: endpointConfig.name,
+                                                apiAuthProviderFactory: apiAuthProviderFactory,
+                                                authService: authService)
+        let authConfiguration = try AWSAuthorizationConfiguration.makeConfiguration(authType: authType,
+                                                                       region: endpointConfig.region,
+                                                                       apiKey: endpointConfig.apiKey)
+        try config.addAuthInterceptorsToEndpoint(endpointType: endpointConfig.endpointType,
+                                                 authConfiguration: authConfiguration)
+
+        // retrieve current interceptors and replace auth interceptor
+        let currentInterceptors = interceptorsForEndpoint(named: endpointConfig.name).filter {
+            !isAuthInterceptor($0)
+        }
+        config.interceptors.append(contentsOf: currentInterceptors)
```

Comment:
why do we create a new one, add it to the current? what if currentInterceptors already have one for this `authType`?

Reply:
the current interceptors will be configured with one authorizationType generated by the CLI, so if the developer selects user pool as the default auth type, the current interceptors will always that one, regardless of an multi-auth set up. so we could do something like

```
let currentInterceptors = interceptorsForEndpoint(named: endpointConfig.name).filter {            
 !isAuthInterceptor($0)        
}
if let authInterceptor = currentInterceptors.where(authType).first() { 
 return currentInterceptors
} else {
  // create it
  // return current + created one
}
```


Feel free to try and optimize this or add a TODO about this

Reply:
Goo call :) I'll try to add this optimization as part of the next PR https://github.com/aws-amplify/amplify-ios/pull/1260